Fetch token via certificate (#206)
promregator/promregator: docs/outbound-authentication.md

Promregator is capable of authenticating towards scraping targets with different schemes. The supported types are

| Type identifier        | Description |
|------------------------|-------------|

> | null                   | A null-authentication, i.e. no addition authentication takes place (may become handy in certain configuration cases |

Reviewer: Boy scouting:
```suggestion
| null                   | A null-authentication, i.e. no addition authentication takes place (may become handy in certain configuration cases) |
```

| basic                  | Basic Authentication scheme using plain-text-based Basic Authentication as defined in [RFC2617](https://www.ietf.org/rfc/rfc2617.txt) |
| OAuth2XSUAA            | OAuth2/JWT-based authentication scheme using grant type "Client Credentials" with basic authentication as used for XSUAA servers. **Deprecated**, use OAuth2XSUAABasic instead |
| OAuth2XSUAABasic       | OAuth2/JWT-based authentication scheme using grant type "Client Credentials" with basic authentication as used for XSUAA servers |

> | OAuth2XSUAACertificate | OAuth2/JWT-based authentication scheme using grant type "Client Credentials" witch certificate based authentication as used for XSUAA servers |

Reviewer: ```suggestion
| OAuth2XSUAACertificate | OAuth2/JWT-based authentication scheme using grant type "Client Credentials" with certificate-based authentication as used for XSUAA servers |
```

### Authentication Configuration for OAuth2XSUAA Authentication

The OAuth2XSUAA Authentication type is deprecated and remains present due to backward compatibility. It is replaced by OAuth2XSUAABasic and behaves in the
same way like that authentication type.

### Authentication Configuration for OAuth2XSUAABasic Authentication

> The OAuth2XSUAABasic Authentication scheme allows to set the following additional configuration options:

Reviewer: ```suggestion
The OAuth2XSUAABasic Authentication scheme allows to set the following configuration options:
```

* `client_id` (mandatory): specifies the client identifier which shall be used when authenticating at the OAuth2 server.
* `client_secret` (mandatory): specifies the client secret which shall be used when authenticating at the OAuth2 server.
* `scopes` (optional): specifies the scopes which shall be requested from the OAuth2 server during the call. If not specified, an empty string is assumed, which will suppress a dedicated request of scopes. Usually, OAuth2 servers then provide a JWT, which contains all scopes allowed for the set of credentials provided.

Note that specifying the secret plain-text in the configuration file is highly discouraged due to security reasons. For an alternative solution, please refer to [this page](./passwords-in-config.md).

An example of a authentication configuration using basic authentication looks like this:

``` yaml
type: oauth2xsuaaBasic
oauth2xsuaaBasic:
  tokenServiceURL: https://instance.subdomain.example.org/oauth/token
  client_id: myclientid
  client_secret: mysecret
  scopes: scopea,scopeb
```

### Authentication Configuration for OAuth2XSUAACertificate Authentication

> The OAuth2XSUAACertificate Authentication scheme allows to set the following additional configuration options:

Reviewer: ```suggestion
The OAuth2XSUAACertificate Authentication scheme allows to set the following configuration options:
```

* `tokenServiceCertURL` (mandatory): specifies the URL of the OAuth2 server, which shall be used to retrieve the token.
* `client_id` (mandatory): specifies the client identifier which shall be used when authenticating at the OAuth2 server.
* `client_certificates` (mandatory): specifies the certificate (chain) which shall be used when authenticating at the OAuth2 server.

> * `client_key` (mandatory): specifies the key which corresponds to the certificate which shall be used when authentication at the OAuth2 server.

Reviewer: ```suggestion
* `client_key` (mandatory): specifies the key which shall be used when authentication at the OAuth2 server. The key must fit to the the certificate used above.
```

* `scopes` (optional): specifies the scopes which shall be requested from the OAuth2 server during the call. If not specified, an empty string is assumed, which will suppress a dedicated request of scopes. Usually, OAuth2 servers then provide a JWT, which contains all scopes allowed for the set of credentials provided.

> An example of a authentication configuration using certificate-based authentication looks like this:

Reviewer: ```suggestion
An example of an authentication configuration using certificate-based authentication looks like this:
```

``` yaml
type: oauth2xsuaaCertificate
oauth2xsuaaCertificate:
  tokenServiceCertURL: https://instance.subdomain.example.org/oauth/token
  client_id: myclientid

>   client_certificates: mycert

Reviewer: ```suggestion
  client_certificates: "-----BEGIN CERTIFICATE-----..."
```

>   client_key: mykey

Reviewer: ```suggestion
  client_key: "-----BEGIN RSA PRIVATE KEY-----..."
```

The OAuth2XSUAA Authentication scheme allows to set the following additional configuration options:

* `url` (mandatory for basic authentication): specifies the URL of the OAuth2 server, which shall be used to retrieve the token.

> * `cert_url` (mandatory for certificate based authentication): specifies the URL of the OAuth2 server, which shall be used to retrieve the token.

Reviewer: ```suggestion
* `cert_url` (mandatory for certificate-based authentication): specifies the URL of the OAuth2 server, which shall be used to retrieve the token.
```

* `client_id` (mandatory): specifies the client identifier which shall be used when authenticating at the OAuth2 server.
* `client_secret` (mandatory for basic authentication): specifies the client secret which shall be used when authenticating at the OAuth2 server.

> * `client_certificates` (mandatory for certificate based authentication): specifies the certificate (chain) which shall be used when authenticating at the OAuth2 server.

Reviewer: ```suggestion
* `client_certificates` (mandatory for certificate-based authentication): specifies the certificate (chain) which shall be used when authenticating at the OAuth2 server.
```

> * `client_key` (mandatory for certificate based authentication): specifies the key which corresponds to the certifate which shall be used when authentication at the OAuth2 server.

Reviewer: ```suggestion
* `client_key` (mandatory for certificate-based authentication): specifies the key which corresponds to the certificate which shall be used when authentication at the OAuth2 server.
```

> * `scope` (optional): specifies the scopes which shall be requested from the OAuth2 server during the call. If not specified, an empty string is assumed, which will suppress a dedicated request of scopes. Usually, OAuth2 servers then provide a JWT, which contains all scopes allowed for the set of credentials provided.

Reviewer: ```suggestion
* `scopes` (optional): specifies the scopes which shall be requested from the OAuth2 server during the call. If not specified, an empty string is assumed, which will suppress a dedicated request of scopes. Usually, OAuth2 servers then provide a JWT, which contains all scopes allowed for the set of credentials provided.
```
BTW: Fixing this typo as a by-catch.

``` yaml
type: oauth2xsuaa
oauth2xsuaa:
  url: https://instance.subdomain.example.org
  client_id: myclientid
  client_secret: mysecret
  scopes: scopea,scopeb
```

> An example of a authentication configuration using certificate based authentication looks like this:

Reviewer: ```suggestion
An example of a authentication configuration using certificate-based authentication looks like this:
```
